include/openPMD/IO/InvalidatableFile.hpp

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+
+#include <string>
+#include <memory>
+
+
+namespace openPMD
+{
+    /**
+     *  Wrapper around a shared pointer to:
+     *  * a filename
+     *  * and a boolean indicating whether the file still exists
+     *  The wrapper adds no extra information, but some commodity functions.
+     *  Invariant for any context within which this class shall be used:
+     *  For any valid filename, there is at any time at most one
+     *  such shared pointer (wrapper) known in said context's data structures
+     *  (counting by pointer equality)
+     *  This means, that a file can be invalidated (i.e. deleted or overwritten)
+     *  by simply searching for one instance of the file among all known files and
+     *  invalidating this instance
+     *  A new instance may hence only be created after making sure that there are
+     *  no valid instances in the data structures.
+     */
+    struct InvalidatableFile
+    {
+        explicit InvalidatableFile( std::string s );
+
+
+        InvalidatableFile( ) = default;
+
+
+        struct FileState
+        {
+            explicit FileState( std::string s );
+
+            std::string name;
+            bool valid = true;
+        };
+
+        std::shared_ptr< FileState > fileState;
+
+
+        void invalidate( );
+
+
+        bool valid( ) const;
+
+
+        InvalidatableFile & operator=( std::string s );
+
+
+        bool operator==( InvalidatableFile const & f ) const;
+
+
+        std::string & operator*( ) const;
+
+
+        std::string * operator->( ) const;
+
+
+        explicit operator bool( ) const;
+    };
+}
+
+namespace std
+{
+    template< >
+    struct hash< openPMD::InvalidatableFile >
+    {
+        typedef openPMD::InvalidatableFile argument_type;
```

Comment:
please use `using` in new code for `typedef`s :)

Reply:
fixed